src/Plugin/RulesExpression/RulesLoop.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\rules\Plugin\RulesExpression\RulesLoop.
+ */
+
+namespace Drupal\rules\Plugin\RulesExpression;
+
+use Drupal\Core\TypedData\ListDataDefinitionInterface;
+use Drupal\rules\Engine\ActionExpressionContainer;
+use Drupal\rules\Engine\ExecutionMetadataStateInterface;
+use Drupal\rules\Engine\ExecutionStateInterface;
+use Drupal\rules\Engine\IntegrityViolationList;
+
+/**
+ * Holds a set of actions that are executed over the iteration of a list.
+ *
+ * @RulesExpression(
+ *   id = "rules_loop",
+ *   label = @Translation("Action set"),
+ *   form_class = "\Drupal\rules\Form\Expression\ActionSetForm"
+ * )
+ */
+class RulesLoop extends ActionExpressionContainer {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function executeWithState(ExecutionStateInterface $state) {
+    $list_data = $state->getVariable($this->configuration['list']);
+    // Use a configured list item variable name, otherwise fall back to just
+    // 'list_item' as variable name.
+    $list_item_name = isset($this->configuration['list_item']) ? $this->configuration['list_item'] : 'list_item';
+
+    foreach ($list_data as $item) {
+      $state->setVariableData($list_item_name, $item);
+      foreach ($this->actions as $action) {
+        $action->executeWithState($state);
+      }
+    }
+    // After the loop the list item is out of scope and cannot be used by any
+    // following actions.
+    $state->deleteVariable($list_item_name);
```

Comment:
In d7 we cloned the state instead so we can leverage php scope handling for our scope. This fixes having a list item variable the overwrites a variable in the parent scope and makes the parent variable available again afterwards. Not sure we really support that, throwing an integrity exception in this case should be ok imo.
I think the d7 approach has some troubles with (open issues) of saving changes on the interim variables, so explictly removing them again + doing the integritycheck seems like a good approach to me. However "deleteVariable" sounds like it's being really deleted, can we call it remove? Also I think it would make sense to have a $save_changes parameter which defaults to TRUE.